src/FoldersView/FolderSourceItem.vala

```diff
@@ -98,5 +117,76 @@ public class Mail.FolderSourceItem : Mail.SourceList.ExpandableItem {
                 is_special_folder = false;
                 break;
         }
+
+        if (!is_special_folder) {
+            editable = true;
+            edited.connect (rename);
+        }
+    }
+
+    private async void refresh () {
+        var offlinestore = (Camel.Store)account.service;
+        try {
+            var folder = yield offlinestore.get_folder (full_name, 0, GLib.Priority.DEFAULT, cancellable);
+            yield folder.refresh_info (GLib.Priority.DEFAULT, cancellable);
+        } catch (Error e) {
+            critical (e.message);
+        }
+    }
+
+    private async void rename (string new_name) {
+        if ("/" in new_name) {
+            if (name == old_name) {
+                notify["name"].connect (() => { name = old_name; });
+            } else {
+                name = old_name;
+            }
+
+            MainWindow.notify_error (
+                _("Unable to rename folder “%s”: Folder names cannot contain “/”").printf (name)
+            );
+
+            return;
+        }
+
+        string[] split_full_name = full_name.split_set ("/");
+        split_full_name[split_full_name.length - 1] = new_name;
+        var new_full_name = string.joinv ("/", split_full_name);
+
+        var offlinestore = (Camel.Store)account.service;
+
+        Camel.FolderInfo? folder_info = null;
+        try {
+            folder_info = yield offlinestore.get_folder_info (new_full_name, FAST, GLib.Priority.DEFAULT, cancellable);
+        } catch (Error e) {
+            warning (e.message);
+        }
+
+        if (null != folder_info) {
+            if (name == old_name) {
+                notify["name"].connect (() => { name = old_name; });
```

Comment:
If the new name is the same as the old, meaning a person decides to not rename, shouldn't it just return? In testing, it feels a bit odd to get an error from backing out of the rename

Reply:
Yeah that's definitely unwanted. Last commit fixes that and some more issues you might have been having